backend/module/chaos/serverexperimentation/xds/ecds.go

```diff
@@ -0,0 +1,302 @@
+	} else {
+		faultFilterName = faultFilterConfigNameForInternalFault
+
+		// We match against the EnvoyDownstreamServiceCluster header to only apply these faults to the specific downstream cluster
+		faultFilter.Headers = []*gcpRoute.HeaderMatcher{
+			{
+				Name: envoyDownstreamServiceClusterHeader,
+				HeaderMatchSpecifier: &gcpRoute.HeaderMatcher_ExactMatch{
+					ExactMatch: downstreamCluster,
+				},
+			},
+		}
+	}
+
+	logger.Debugw("ECDS Fault filter", "faultFilter", faultFilter, "upstreamcluster", upstreamCluster)
+
+	serializedFaultFilter, err := proto.Marshal(faultFilter)
+	if err != nil {
+		logger.Warnw("Unable to unmarshal fault filter", "faultFilter", faultFilter)
+		return []gcpTypes.ResourceWithTtl{}
+	}
+
+	return []gcpTypes.ResourceWithTtl{{
+		Resource: &gcpCoreV3.TypedExtensionConfig{
+			Name: faultFilterName,
+			TypedConfig: &any.Any{
+				TypeUrl: faultFilterTypeURL,
+				Value:   serializedFaultFilter,
+			},
+		},
+		Ttl: ttl,
+	}}
+}
+
+func generateEmptyECDSResource(cluster string, ecdsConfig *ECDSConfig, logger *zap.SugaredLogger) []gcpTypes.ResourceWithTtl {
+	// cache serialized
+	if SerializedDefaultFaultFilter == nil {
+		faultFilter := &gcpFilterFault.HTTPFault{
+			Delay: DefaultDelayFaultConfig,
+			Abort: DefaultAbortFaultConfig,
+
+			// override runtimes so that default runtime is not used.
+			DelayPercentRuntime:    delayPercentRuntime,
+			DelayDurationRuntime:   delayDurationRuntime,
+			AbortHttpStatusRuntime: abortHttpStatusRuntime,
+			AbortPercentRuntime:    abortPercentRuntime,
+		}
+
+		SerializedDefaultFaultFilter, _ = proto.Marshal(faultFilter)
+	}
+
+	if _, exists := ecdsConfig.ecdsResourceMap.requestedResourcesMap[cluster]; !exists {
+		logger.Debugw("Cluster not found in ECDS resource map", "cluster", cluster)
+		return []gcpTypes.ResourceWithTtl{}
+	}
+
+	resourceNames := ecdsConfig.ecdsResourceMap.getResourcesFromCluster(cluster)
+
+	var resources []gcpTypes.ResourceWithTtl
+	for _, resourceName := range resourceNames {
+		resource := gcpTypes.ResourceWithTtl{
+			Resource: &gcpCoreV3.TypedExtensionConfig{
+				Name: resourceName,
+				TypedConfig: &any.Any{
+					TypeUrl: faultFilterTypeURL,
+					Value:   SerializedDefaultFaultFilter,
+				},
+			},
+			Ttl: nil,
+		}
+		resources = append(resources, resource)
+	}
+
+	return resources
+}
+
+func createAbortDelayConfig(httpFaultConfig *serverexperimentation.HTTPFaultConfig, downstreamCluster string) (bool, *gcpFilterFault.FaultAbort, *gcpFilterCommon.FaultDelay, error) {
+	var isExternalFault bool
+	var abort *gcpFilterFault.FaultAbort
+	var delay *gcpFilterCommon.FaultDelay
+
+	switch httpFaultConfig.GetFault().(type) {
+	case *serverexperimentation.HTTPFaultConfig_AbortFault:
+		switch httpFaultConfig.GetFaultTargeting().GetEnforcer().(type) {
+		case *serverexperimentation.FaultTargeting_DownstreamEnforcing:
+			// Abort External Fault
+			isExternalFault = true
```

Comment:
It seems a bit odd for someone not familiar with the lyft use case to see external used here since the API just talks about downstream vs upstream enforcing. Maybe be consistent and call this "isEgressFault" or something?